### PR TITLE
[Fluid] Add pressure gradient relaxation for fractional step

### PR DESCRIPTION
**Pull request by sunethwarna**
**Description**
This adds additional pressure gradient relaxation for fractional step to acheive steady state solutions. It is formulated based on work done by 

A. R. Firoozjaee and M. H. Afshar, Steady-state solution of incompressible navier–stokes equations using discrete least-squares meshless method, International Journal for Numerical Methods in Fluids, 67 (2011), pp. 369–382.

**Changelog**
- Added `FS_PRESSURE_GRADIENT_RELAXATION_FACTOR`
- Added pressure gradient modification to `FractionalStepStrategy`
- Added pressure gradient modification to `FractionalStep` element

### CODE REVIEW

**Comment by philbucher** [comment]
is this a behavior change?

**Comment by sunethwarna** [comment]
@philbucher No, It is not a behaviour change. This relaxation is optional.

**Comment by RiccardoRossi** [comment]
@sunethwarna wouldn't it make sense to actually change the nodal values? that could be done at a lower cost in the strategy, and would require no change in the element

**Comment by sunethwarna** [comment]
@RiccardoRossi If i am to use nodal update values with then there are two problems.
1. Then the final `PRESSURE` value will be multiplied with the `eta` value, which will be a problem for transient simulations
2. It will change the formulation specially concerning the stabilization, since I am using stabilization with actual pressure (without `eta` modification in order not to change the strong form which the stabilization residual is based on.

I did try to reformulate this to do everything in the strategy, but unfrtunately I couldn't do the same without modifying element :/

**Review by RiccardoRossi — COMMENTED** [review]
can t u change it before step0 in the strwtrgy and then change it back, for example just after the step?

**Comment by sunethwarna** [comment]
Hi @RiccardoRossi. For further clarification I include the followings and modifications I introduce.

First step is to find intermediate fractional velocity with `FRACTIONAL_STEP=1`
LHS matrix calculatd at `CalculateLocalFractionalVelocitySystem` in `FractionalStep` element:
image
RHS vector calculatd at `CalculateLocalFractionalVelocitySystem` in `FractionalStep` element: 
image

In there T_0 and T^* are time dependent constant and a vector. M^{ab}_{ij} is the mass matrix. P^{mom}_i is the momentum projections 3D vector, P^{mass} is the mass projection scalar. I didn't expand them because they are not required for this discussion.

Afterwards Projection values are updated. Then Pressure is calculated with `FRACTIONAL_STEP=5`
LHS matrix calculated at `CalculateLocalPressureSystem` in `FractionalStep` element:
image
RHS vector calculated at `CalculateLocalPressureSystem` in `FractionalStep` element:
image

The final velocity update at `CalculateEndOfStepVelocity` with `FRACTIONAL_STEP=6`
image

The above equations represent current `FractionalStep` implementation without pressure gradient relaxation. What I am proposing is following modifications to existing implementation.

In `CalculateLocalFractionalVelocitySystem` with `FRACTIONAL_STEP=1` RHS:
image
In `CalculateLocalPressureSystem` with `FRACTIONAL_STEP=5` RHS:
image
Above equation for pressure solve is derrived from the following:
image

At the end, updating velocities with modified pressure gradient relaxation as following at `CalculateEndOfStepVelocity` with `FRACTIONAL_STEP=6`
image

Current proposal adds the modifications at `FractionalStepStrategy` as well as `FractionalStep` element. Both reads `FS_PRESSURE_GRADIENT_RELAXATION_FACTOR` from process info which is by default `1.0` so any of the proposed changes do not affect current behaviour.

**If reading from process info in elements adds significant amount of overhead, then I can do the following:**

I can as you suggest multiply `PRESSURE`  with `eta`, so we don't have to modify `CalculateLocalFractionalVelocitySystem` method in `FractionalStep` element. But I still have to modify `CalculateLocalPressureSystem` to use `PRESSURE_OLD_IT` to achieve following modification:
image

This method will introduce minimal changes to `FractionalStep` element, and it does not have to read `eta` value from `FS_PRESSURE_GRADIENT_RELAXATION_FACTOR` in process info. I think this way will be much better in terms of the performance as well. 

Could you please have a look at this and let me know what do you think?

Thanks :)

File: applications/FluidDynamicsApplication/custom_strategies/strategies/fractional_step_strategy.h [hunk 1]
```diff
@@ -956,6 +959,17 @@ class FractionalStepStrategy : public SolvingStrategy<TSparseSpace, TDenseSpace,
 
         mReformDofSet = rSolverConfig.GetReformDofSet();
 
+        auto& r_process_info = BaseType::GetModelPart().GetProcessInfo();
+        if (r_process_info.Has(FS_PRESSURE_GRADIENT_RELAXATION_FACTOR)) {
+            mPressureGradientRelaxationFactor = r_process_info[FS_PRESSURE_GRADIENT_RELAXATION_FACTOR];
+            KRATOS_INFO(this->Info()) << "Using fractional step strategy with "
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_strategies/strategies/fractional_step_strategy.h]:
```suggestion
            KRATOS_INFO("FractionalStepStrategy") << "Using fractional step strategy with "
```
We discussed about using `Info` to print the names in another PR. We agreed that this shouldn't be done as info might contain more information than the name.

File: applications/FluidDynamicsApplication/custom_utilities/fractional_step_settings.h [hunk 2]
```diff
@@ -103,6 +103,7 @@ class FractionalStepSettings: public SolverSettings<TSparseSpace,TDenseSpace,TLi
     ///@name Access
     ///@{
 
+
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_utilities/fractional_step_settings.h]: I think this has been accidentally committed.

**Review by rubenzorrilla — COMMENTED** [review]
1st of all, many thanks for such a clarifying explanation.

IMO the current element can be modified provided that the default behaviour is kept (basically what @sunethwarna did). In any case, if this is a no-go we can create a derived FS element that only implement the two new contributions.

I'd like to add that I don't think the process info access is a drama (we're already accessing to compute the stabilization constants).

File: applications/FluidDynamicsApplication/custom_utilities/fractional_step_settings.h (hunk 2 above)
Comment (sunethwarna) [inline applications/FluidDynamicsApplication/custom_utilities/fractional_step_settings.h]:
```suggestion
```

Comment (sunethwarna) [reply applications/FluidDynamicsApplication/custom_utilities/fractional_step_settings.h]: sorry, it was added accidentally. And now removed :)

**Review by rubenzorrilla — APPROVED** [review]
After a brief discussion with @RiccardoRossi we agree that the changes in the element are required. Approving on behalf of him.

**Comment by sunethwarna** [comment]
thanks @rubenzorrilla and @RiccardoRossi  :)

**Pull request merged**